corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +16,269 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        for key in dict(data, **uncategorized_data):
+            value = self.user.metadata[key]
+            if value is None or value == '':
+                self.user.pop_metadata(key)
+
+    def update_language(self, language):
+        self.user.language = language
+        self.logger.add_changes({'language': language})
+
+    def update_email(self, email):
+        self.user.email = email.lower()
+        self.logger.add_changes({'email': self.user.email})
+
+    def update_status(self, is_active):
+        self.user.is_active = is_active
+        self.logger.add_changes({'is_active': is_active})
+
+    def update_locations(self, location_codes, domain_info):
+        from corehq.apps.user_importer.importer import (
+            check_modified_user_loc,
+            find_location_id,
+            get_location_from_site_code
+        )
+
+        location_ids = find_location_id(location_codes, domain_info.location_cache)
+        users_current_primary_location_id = self.user.location_id
+        locations_updated, primary_loc_removed = check_modified_user_loc(location_ids,
+                                                                         self.user.location_id,
+                                                                         self.user.assigned_location_ids)
+        if primary_loc_removed:
+            self.user.unset_location(commit=False)
+        if locations_updated:
+            self.user.reset_locations(location_ids, commit=False)
+            self.logger.add_changes({'assigned_location_ids': location_ids})
+            if location_ids:
+                location_names = [get_location_from_site_code(code, domain_info.location_cache).name
+                                  for code in location_codes]
+                self.logger.add_info(_("Assigned locations: {location_names}").format(
+                    location_names=location_names
+                ))
+
+        # log this after assigned locations are updated, which can re-set primary location
+        if self.user.location_id != users_current_primary_location_id:
+            self.logger.add_changes({'location_id': self.user.location_id})
+            if self.user.location_id:
+                users_updated_primary_location_name = get_user_primary_location_name(self.user, self.user_domain)
+                self.logger.add_info(
+                    _("Primary location: {primary_location_name}").format(
+                        primary_location_name=users_updated_primary_location_name
+                    ))
+
+
+def _fmt_phone(phone_number):
+    if phone_number and not isinstance(phone_number, str):
+        phone_number = str(int(phone_number))
+    return phone_number.lstrip("+")
+
+
+class WebUserImporter(BaseUserImporter):
+    def add_to_domain(self, role_qualified_id, location_id):
+        self.user.add_as_web_user(self.user_domain, role=role_qualified_id, location_id=location_id)
+        self.logger.add_change_message(_("Added as web user"))
```

Comment:
This could use more detail: `"Added user to domain '{domain}' with role '{role}' and location '{location}'"`

Reply:
I corrected the documentation added. Role and location are assigned only if web user is already a member of the domain, so that would not happen here.

Also generally, Role and location information will get added when and if those are assigned.

Just thought,
My sense with these messages is that it should avoid implementation context and should be specific to what changed. So not saying that `User was added with foo role and bar location`, but saying, `User was added. Role foo. Location Bar` to have a simpler and common way of adding them from all places.

Reply:
Documentation updated https://github.com/dimagi/commcare-hq/pull/29935/commits/a6dabaccbe0b14804b550c86b296b3d6ceea7957

Reply:
I don't quite follow, if the user is already a member of the domain and the role and location are handled elsewhere they why are we calling this method which does all 3 of those things?

My point is that "Added as web user" is not going to mean anything to a user reading the logs and it would be better to give a message that a user is going to understand.

Reply:
> I don't quite follow, if the user is already a member of the domain and the role and location are handled elsewhere they why are we calling this method which does all 3 of those things?

Ah, I see. You are right. We are adding user along with a role and possibly with a location. So we should also track that here.

https://github.com/dimagi/commcare-hq/pull/29935/commits/29d0abf4f7291a23e10dd012c46f078a395245d7

> My point is that "Added as web user" is not going to mean anything to a user reading the logs and it would be better to give a message that a user is going to understand.

Just generally, its referring to that the user was added as a member to the domain. If we were to find out when a user started getting access to a domain, this would be the log to look at. Does that make sense?

Reply:
also, as a follow up https://github.com/dimagi/commcare-hq/pull/29935/commits/9fa95b0ab6b284a953702277d51520d248b994ff

Reply:
Yes it does, I'm just thinking about how this will show up in the report. Maybe we could re-word it to "Added as web user to this domain"? Assuming that the context of the report is the domain that the user was added to?

I'm also thinking that if you were just looking at the logs for the web user (across all domains) you'd need to consider which domain the log was created for in order to know which domain they were added to.

Reply:
>  Maybe we could re-word it to "Added as web user to this domain"? Assuming that the context of the report is the domain that the user was added to?

Adding domain name in this itself will be useful when we have multi domain report. Users for a master domain, looking at report for users in all linked domains. 

For example:
When filtering on a master domain,
User should see changes done for users in this domain and the linked domains,
by users belonging to this domain and the linked domains,
Where logs would look like,
User    |              Message
User1  | Added as web user to cape-town. Role: SSE
User1  | Added as web user to boston. Role: Read-Only. Primary Location: DimagiOffice

> I'm also thinking that if you were just looking at the logs for the web user (across all domains) you'd need to consider which domain the log was created for in order to know which domain they were added to.

I see your point. Similar to "by_domain", we should also have a "for_domain". 
I was thinking but don't see the use case yet. When anything is updated on the user that is domain specific, like location, role, they all have a "domain" attribute on it, I believe, mostly through domain membership. Can't totally rule out the need for adding "for_domain" in future.